README.md
# kv-logger ![Node.js CI](https://github.com/rocwind/kv-logger/workflows/Node.js%20CI/badge.svg)

---

a tiny logger that supports log with k-v params.

## Install

`npm install --save kv-logger`

## Usage Example

```
// es module
import { logger } from 'kv-logger';
// or common module
const { logger } = require('kv-logger');

logger.info('info message');
logger.warn('warning message', {
    user_id: 123,
});

const loggerWithContext = logger.bindContext({ match_id: 12345 });
loggerWithContext.error('error message with context');
```

check out more examples at: [example.js](example/example.js)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.5.3](https://github.com/rocwind/kv-logger/compare/v0.5.2...v0.5.3) (2022-02-18)


### Bug Fixes

* export Logger class ([673728b](https://github.com/rocwind/kv-logger/commit/673728bfbf4ab3994fffbcb7210875dfb33e0f0d))

### [0.5.2](https://github.com/rocwind/kv-logger/compare/v0.5.1...v0.5.2) (2022-02-11)


### Bug Fixes

* correct LogMessage type ([039d0a4](https://github.com/rocwind/kv-logger/commit/039d0a45c1518c5952fbe52a0d371f3ee36b4e1b))

### [0.5.1](https://github.com/rocwind/kv-logger/compare/v0.5.0...v0.5.1) (2022-02-11)


### Bug Fixes

* add withLogLevelFilter, deprecate withLogLevel ([89592ab](https://github.com/rocwind/kv-logger/commit/89592abe11e458394e52da215f7de7997c8c3552))
* setLogTransports() supports set single transport ([3062f2f](https://github.com/rocwind/kv-logger/commit/3062f2ff14591aa11d7ee24e249934312fa02e92))

## [0.5.0](https://github.com/rocwind/kv-logger/compare/v0.4.1...v0.5.0) (2022-02-11)


### ⚠ BREAKING CHANGES

* change LogLevel to string enum, rename ConsoleTransport to ConsoleLogTransport

### Features

* support withLogScope to set global log context ([fe7728f](https://github.com/rocwind/kv-logger/commit/fe7728f39b49f433b3559cb5d723ee80da04aafa))

### [0.4.1](https://github.com/rocwind/kv-logger/compare/v0.4.0...v0.4.1) (2022-02-08)


### Bug Fixes

* LogLevelFilter takes string level type as parameter ([f3523a7](https://github.com/rocwind/kv-logger/commit/f3523a71619fb985d064df20e6c267912d2d8d12))

## [0.4.0](https://github.com/rocwind/kv-logger/compare/v0.3.0...v0.4.0) (2022-02-07)


### ⚠ BREAKING CHANGES

* remove the default export of logger instance

### Features

* deprecate enums ConsoleFormat and LogLevel, use string values in api ([2597aa8](https://github.com/rocwind/kv-logger/commit/2597aa8838bd573c837614eed27834f9e7bf88a6))


### Bug Fixes

* remove the default export of logger instance ([3214b86](https://github.com/rocwind/kv-logger/commit/3214b86abad67e2b85f1b6c2d14aeda4d6d71b0f))
* use formated date time for console text log ([4de1132](https://github.com/rocwind/kv-logger/commit/4de1132d93117eda44bce77aa4a7783e148ac9aa))

## [0.3.0](https://github.com/rocwind/kv-logger/compare/v0.2.2...v0.3.0) (2020-01-06)


### ⚠ BREAKING CHANGES

* refactor LogTransport interface

### Features

* refactor LogTransport interface ([26639a3](https://github.com/rocwind/kv-logger/commit/26639a36743479396a5c6ea70611ac35c25f92a7))

### [0.2.2](https://github.com/rocwind/kv-logger/compare/v0.2.1...v0.2.2) (2019-12-14)


### Features

* add logger to default export ([a4430fb](https://github.com/rocwind/kv-logger/commit/a4430fb2429b2f2b331056655588cef1e9014ba1))


### Bug Fixes

* setLogTransports() make a copy of input array ([2e8b385](https://github.com/rocwind/kv-logger/commit/2e8b385abacdc2d13eadf57f8f852d41c573580e))

### [0.2.1](https://github.com/rocwind/kv-logger/compare/v0.2.0...v0.2.1) (2019-12-10)


### Features

* add log() method ([6bd8c6f](https://github.com/rocwind/kv-logger/commit/6bd8c6f4b4c713ecbe5a4fce0696a449c1b50864))

## [0.2.0](https://github.com/rocwind/kv-logger/compare/v0.1.0...v0.2.0) (2019-12-10)


### ⚠ BREAKING CHANGES

* global setConfig() is removed, LogTransport interface is updated

### Features

* refactor the transports to support customized format ([07a8088](https://github.com/rocwind/kv-logger/commit/07a8088b51196d91ad2420c643050acd5cb4c64a))

## 0.1.0 (2019-12-09)


### Features

* initial commit ([ce8f54e](https://github.com/rocwind/kv-logger/commit/ce8f54e26607081151ebe2d732421c930854e250))


src/__tests__/test.js
const {
    logger,
    setLogTransports,
    ConsoleFormat,
    ConsoleTransport,
    LogLevel,
    LogLevelFilter,
} = require('../../lib/bundle.js');

// config transport to console+ text format, filter log level by info and above
setLogTransports([
    new LogLevelFilter(new ConsoleTransport(ConsoleFormat.Text), LogLevel.Info),
]);

// plain text
logger.debug("you won't see this debug log as current level is info");
logger.warn('and this one is visible');

// k-v by object
logger.error({
    user_id: 12345,
    match_id: 12315,
    msg: 'connection lost.',
});

// msg + k-v object
logger.info('reconnecting...', {
    user_id: 12345,
    match_id: 12315,
});

// ooops, there is no msg key in object
logger.info({
    text: 'forgot the msg key',
});

// the error object
logger.error(new Error('this is an error'));

// now let's bind k-v context to a logger
var loggerWithContext = logger.bindContext({ match_id: 12425, user_id: 12315 });
loggerWithContext.info('log with context');

// 2nd level of context binding
var loggerWithSubContext = loggerWithContext.bindContext({ worker_id: 'a' });
loggerWithSubContext.info('log with sub context');

// config to json output and no log level filter
setLogTransports([new ConsoleTransport(ConsoleFormat.JSON)]);

logger.debug('debug message is visible now');

logger.info({
    msg: 'info log with k-v in json',
    user_id: 12451,
});


src/scope.ts
const scopeStack: Record<string, unknown>[] = [];

/**
 * current scope context
 * @returns
 */
export function getLogScope(): Record<string, unknown> {
    return scopeStack[scopeStack.length - 1];
}

/**
 * push a new scope to the stack
 * @param context
 */
export function pushLogScope(context: Record<string, unknown>): void {
    scopeStack.push(Object.assign({}, getLogScope(), context));
}

/**
 * pop the current scope context
 */
export function popLogScope() {
    if (scopeStack.length > 1) {
        scopeStack.pop();
    }
}

/**
 * execute callback with given scope context
 * @param context
 * @param callback
 */
export function withLogScope(
    context: Record<string, unknown>,
    callback: () => void
) {
    pushLogScope(context);
    try {
        callback();
    } finally {
        popLogScope();
    }
}


src/format.ts
const leftPadNumberStr = (num: number, digest = 2): string => {
    let bound = 10;
    let prefix = '';
    for (let i = 1; i < digest; i++) {
        if (num < bound) {
            prefix += '0';
        }
        bound *= 10;
    }
    return prefix + num;
};

/**
 * YYYY-MM-DD hh:mm:ss.sss
 * @param timestamp
 */
export const formatTime = (timestamp: number): string => {
    const date = new Date(timestamp);
    return (
        date.getFullYear() +
        '-' +
        leftPadNumberStr(date.getMonth() + 1) +
        '-' +
        leftPadNumberStr(date.getDate()) +
        ' ' +
        leftPadNumberStr(date.getHours()) +
        ':' +
        leftPadNumberStr(date.getMinutes()) +
        ':' +
        leftPadNumberStr(date.getSeconds()) +
        '.' +
        leftPadNumberStr(date.getMilliseconds(), 3)
    );
};


src/logger.ts
import { getLogScope } from './scope';
import { ConsoleLogTransport } from './transports';
import { LogLevel, LogTransport, Log, LogMessage } from './types';

/**
 * default transport is console + text format
 */
let transports: LogTransport[] = [new ConsoleLogTransport('text')];

/**
 * configure log transports
 * @param entries
 */
export function setLogTransports(entries: LogTransport[] | LogTransport): void {
    if (Array.isArray(entries)) {
        transports = entries.slice();
    } else {
        transports = [entries];
    }
}

function composeLog(
    level: LogLevel,
    msg: LogMessage,
    params?: Record<string, unknown>,
    context?: Record<string, any>
): Log {
    // generate the log object keys with { level, time, msg, ... }
    const log: Log = Object.assign(
        { level, time: 0, msg: '' },
        getLogScope(),
        context,
        params
    );
    if (typeof msg === 'string') {
        log.msg = msg;
    } else if (msg instanceof Error) {
        log.msg = msg.message;
        log.name = msg.name;
        log.stack = msg.stack;
    } else {
        Object.assign(log, msg);
    }

    log.level = level;
    log.time = log.time || Date.now();

    return log;
}

export class Logger {
    constructor(private context?: Record<string, unknown>) {}

    /**
     * create a child logger with context
     * @param context
     * @returns
     */
    bindContext(context: Record<string, unknown>) {
        return new Logger(Object.assign({}, this.context, context));
    }

    /**
     * write log
     * @param level
     * @param msg
     * @param params
     */
    log(level: LogLevel, msg: LogMessage, params?: Record<string, unknown>) {
        const log = composeLog(level, msg, params, this.context);
        transports.forEach((transport) => transport.write(log));
    }

    /**
     * write debug log
     * @param msg
     * @param params
     */
    debug(msg: LogMessage, params?: Record<string, unknown>) {
        this.log('debug', msg, params);
    }

    /**
     * write info log
     * @param msg
     * @param params
     */
    info(msg: LogMessage, params?: Record<string, unknown>) {
        this.log('info', msg, params);
    }

    /**
     * write warn log
     * @param msg
     * @param params
     */
    warn(msg: LogMessage, params?: Record<string, unknown>) {
        this.log('warn', msg, params);
    }

    /**
     * write error log
     * @param msg
     * @param params
     */
    error(msg: LogMessage, params?: Record<string, unknown>) {
        this.log('error', msg, params);
    }
}

/**
 * create a logger instance
 * @param context the context of the logger
 * @returns
 */
export function createLogger(context?: Record<string, unknown>) {
    return new Logger(context);
}

/**
 * default logger instance
 */
export const logger = createLogger();


src/transports.ts
import { Log, LogTransport, LogLevel } from './types';
import { formatTime } from './format';

const logPriorityByLevel: Record<LogLevel, number> = {
    debug: 1,
    info: 2,
    warn: 3,
    error: 4,
};

/**
 * Filter helper class for filter logs by level
 */
export class LogLevelFilter implements LogTransport {
    constructor(private transport: LogTransport, private level: LogLevel) {}

    write(log: Log) {
        const { level } = log;
        if (logPriorityByLevel[level] < logPriorityByLevel[this.level]) {
            return;
        }
        this.transport.write(log);
    }
}

/**
 * create a log transport with log level filter
 * e.g. withLogLevelFilter('info')(new ConsoleTransport('text'))
 * @param level
 * @returns
 */
export function withLogLevelFilter(
    level: LogLevel
): (transport: LogTransport) => LogTransport {
    return (transport: LogTransport) => new LogLevelFilter(transport, level);
}

/**
 * @deprecated please use withLogLevelFilter()
 */
export const withLogLevel = withLogLevelFilter;

/**
 * Default console log transport, output logs directly to console
 */
export class ConsoleLogTransport implements LogTransport {
    constructor(private format: 'text' | 'json') {}

    write(log: Log) {
        // format time to readable string for log
        const data = Object.assign({}, log, { time: formatTime(log.time) });
        let text: string;
        switch (this.format) {
            case 'json':
                text = JSON.stringify(data);
                break;
            case 'text':
            default:
                text = Object.keys(data)
                    .map((key) => key + '=' + data[key])
                    .join(', ');
                break;
        }

        console[log.level](text);
    }
}


example/example.js
const {
    logger,
    setLogTransports,
    ConsoleLogTransport,
    withLogLevelFilter,
    withLogScope,
} = require('../lib/bundle.js');

// config transport to console+ text format, filter log level by info and above
setLogTransports([withLogLevelFilter('info')(new ConsoleLogTransport('text'))]);

// plain text
logger.debug("you won't see this debug log as current level is info");
logger.warn('and this one is visible');

// k-v by object
logger.error({
    user_id: 12345,
    match_id: 12315,
    msg: 'connection lost.',
});

// msg + k-v object
logger.info('reconnecting...', {
    user_id: 12345,
    match_id: 12315,
});

// ooops, there is no msg key in object
logger.info({
    text: 'forgot the msg key',
});

// the error object
logger.error(new Error('this is an error'));

// now let's bind k-v context to a logger
const loggerWithContext = logger.bindContext({
    match_id: 12425,
    user_id: 12315,
});
loggerWithContext.info('log with context');

// 2nd level of context binding
const loggerWithSubContext = loggerWithContext.bindContext({ worker_id: 'a' });
loggerWithSubContext.info('log with sub context');

// config to json output and no log level filter
setLogTransports([new ConsoleLogTransport('json')]);

logger.debug('debug message is visible now');

logger.info({
    msg: 'info log with k-v in json',
    user_id: 12451,
});

// config scope context for log
withLogScope(
    {
        scope: 1,
    },
    () => {
        logger.info('log in scope');

        withLogScope({ scope: '1-1' }, () => {
            logger.info('log in inner scope');
        });

        logger.info('log 2 in scope');
    }
);
